src/mono/browser/runtime/heapshot/index.ts

```diff
@@ -0,0 +1,328 @@
+/* eslint-disable no-console */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import ProductVersion from "consts:productVersion";
+import cwraps from "../cwraps";
+import { loaderHelpers } from "../globals";
+import { mono_log_info } from "../logging";
+import { getU32 } from "../memory";
+import { utf8ToString } from "../strings";
+import { BlobBuilder } from "../jiterpreter-support";
+import { enumerateProxies } from "../gc-handles";
+import { JiterpCounter, JiterpCounterNames } from "../jiterpreter-enums";
+import { getCounter } from "../jiterpreter-support";
+import type { VoidPtr, ManagedPointer, CharPtr } from "../types/emscripten";
+
+const packetBuilderCapacity = 65536;
+const packetFlushThreshold = 32768;
+// FIXME: It would be ideal if we didn't have to actually retain the whole string
+const stringTable = new Map<string, number>();
+const incompletePackets = new Map<string, BlobBuilder>();
+const formatVersion = 1;
+
+const nameFromKind = ["unknown", "def", "gtd", "ginst", "gparam", "array", "pointer"];
+const handleTypes = ["weak", "weak_track", "normal", "pinned", "weak_fields"];
+
+let totalObjects = 0, totalRefs = 0, totalClasses = 0, totalAssemblies = 0, totalRoots = 0;
+let mostRecentObjectPointer : ManagedPointer = <any>0;
+let heapshotStartedWhen = performance.now();
+
+const tabId = (Math.random() * 100000).toFixed(0);
+const globalChannel : BroadcastChannel | null = globalThis["BroadcastChannel"] ? new BroadcastChannel(".NET Runtime Diagnostics") : null,
```

Comment:
I'm also worried about this becoming attack surface. Let's make it opt-in.

Reply:
It only works for content on the same origin. How would it be used for attacks?

I'm open to opt-in behind a flag or something, but I don't know how to read runtime options from typescript this early in startup.

Reply:
Maybe it could be only listening on the channel after the user opens browser console and enters
```
globalThis.getDotnetRuntime(0).enable_diagnostic_channel()
```

And maybe we could get rid of the whole complexity of HTML deployment if the whole experience was triggered just via
```
globalThis.getDotnetRuntime(0).download_memory_dump()
globalThis.getDotnetRuntime(0).log_memory_stats()
```

I would prefer that.